### PR TITLE
feat: add an ability to edit user name gf-89

### PR DESCRIPTION
**Pull request by perekotypole**
- added profile page with user info edit form
- disabled button and input components
- implemented ability to update user data in backend
- authenticated user data refreshes on updating
- implemented notification on successful data update

image

### CODE REVIEW

File: apps/frontend/src/libs/components/popup/popup.tsx
[the remark this answers is not in the document]
Comment (perekotypole) [reply apps/frontend/src/libs/components/popup/popup.tsx:1]: Yeah, I know... I used the old code, and after was waiting for something more stable

File: apps/frontend/src/libs/helpers/compare-objects/compare-objects.helper.ts [hunk 1]
```diff
@@ -0,0 +1,37 @@
+const compareObjects = <T extends Record<string, unknown>>(
```
[the remark this answers is not in the document]
Comment (perekotypole) [reply apps/frontend/src/libs/helpers/compare-objects/compare-objects.helper.ts:1]: Thanks! I couldn't think of anything more appropriate than `compare`

File: apps/frontend/src/libs/helpers/compare-objects/compare-objects.helper.ts [hunk 2]
```diff
@@ -0,0 +1,37 @@
...
+			return false;
+		}
+	}
+
+	return true;
+};
+
+const isPlainObject = (value: unknown): value is Record<string, unknown> => {
+	return typeof value === "object" && value !== null && !Array.isArray(value);
+};
```
[the remark this answers is not in the document]
Comment (perekotypole) [reply apps/frontend/src/libs/helpers/compare-objects/compare-objects.helper.ts:35]: You are right! I didn't find it in quality criteria, but it makes sense

File: apps/frontend/src/modules/users/slices/users.ts [hunk 3]
```diff
@@ -1,9 +1,10 @@
-import { loadAll } from "./actions.js";
+import { loadAll, updateUser } from "./actions.js";
 import { actions } from "./users.slice.js";
 
 const allActions = {
 	...actions,
 	loadAll,
+	updateUser,
```
Comment (GvoFor) [inline apps/frontend/src/modules/users/slices/users.ts:7]: `updateUser` action will be used like `usersActions.updateUser(...)`, so it's better to rename it to `update`, so we do not violate A5 of QC

File: apps/frontend/src/pages/profile/components/edit-user-form/edit-user-form.tsx [hunk 4]
```diff
@@ -0,0 +1,87 @@
...
+		useAppForm<UserInfoResponseDto>({
+			defaultValues,
+		});
+
+	const [isChanged, setIsChanged] = useState<boolean>(false);
+
+	const watchedValues = watch();
+
+	const handleFormSubmit = useCallback(
+		(event_: React.BaseSyntheticEvent): void => {
```
Comment (GvoFor) [inline apps/frontend/src/pages/profile/components/edit-user-form/edit-user-form.tsx:36]: Why there is underscore `_` ?
```suggestion
		(event: React.BaseSyntheticEvent): void => {
```

File: apps/frontend/src/pages/profile/profile.tsx [hunk 5]
```diff
@@ -0,0 +1,39 @@
...
+		<PageLayout>
+			<div className={styles["profile-content"]}>
+				<h1 className={styles["label"]}>Profile</h1>
+
+				{isLoading ? (
+					<div className={styles["profile-loader"]}>
+						<Loader />
+					</div>
+				) : (
+					authenticatedUser && (
```
Comment (GvoFor) [inline apps/frontend/src/pages/profile/profile.tsx:24]: It doesn't make sense to render the page if authenticatedUser is undefined/null. It's better to add
```ts
if (!authenticatedUser) {
  return <></>;
}
```
right after useAppSelector

BTW `ProtectedRoute` doesn't even allow non-authenticated user to visit this page

File: packages/shared/src/modules/users/libs/types/user-info-request-dto.type.ts [hunk 6]
```diff
@@ -0,0 +1,5 @@
+import { type UserInfoResponseDto } from "./user-info-response-dto.type.js";
+
+type UserInfoRequestDto = UserInfoResponseDto;
```
Comment (GvoFor) [inline packages/shared/src/modules/users/libs/types/user-info-request-dto.type.ts:3]: I think it's better to duplicate type declaration and avoid this assignment in case types might change in the future (and from semantic perspective too)

File: apps/frontend/src/libs/components/button/button.tsx [hunk 7]
```diff
@@ -1,24 +1,36 @@
...
+	className?: string;
+	disabled?: boolean;
 	href?: string | undefined;
 	label: string;
 	type?: "button" | "submit";
 };
 
-const Button = ({ href, label, type = "button" }: Properties): JSX.Element => {
+const Button = ({
+	className = "",
```
Comment (GvoFor) [inline apps/frontend/src/libs/components/button/button.tsx:16]: Adding `className` prop to common components is not good idea. In case you want to control the size of the button, you can do this with wrapper classes. Check this [comment](https://github.com/BinaryStudioAcademy/bsa-2024-gitfit/issues/87#issuecomment-2316856589). Also you can check my [PR](https://github.com/BinaryStudioAcademy/bsa-2024-gitfit/pull/115) where I also modify this component and how I use it in `ProjectCreateForm`

File: apps/frontend/src/modules/users/slices/users.ts (hunk 3 above)
Comment (perekotypole) [reply apps/frontend/src/modules/users/slices/users.ts:7]: Fair enough

File: apps/frontend/src/pages/profile/components/edit-user-form/edit-user-form.tsx (hunk 4 above)
Comment (perekotypole) [reply apps/frontend/src/pages/profile/components/edit-user-form/edit-user-form.tsx:36]: I took the example from sign in/up forms

File: packages/shared/src/modules/users/libs/types/user-info-request-dto.type.ts (hunk 6 above)
Comment (perekotypole) [reply packages/shared/src/modules/users/libs/types/user-info-request-dto.type.ts:3]: I assumed that it would contain absolutely all user data (without passwords and id) and would be reused fot it wherever it was needed.

Comment (liza-veis) [reply packages/shared/src/modules/users/libs/types/user-info-request-dto.type.ts:3]: For request we only need name, so it should be different from response

File: apps/frontend/src/libs/components/button/button.tsx (hunk 7 above)
Comment (perekotypole) [reply apps/frontend/src/libs/components/button/button.tsx:16]: got it

File: packages/shared/src/modules/users/libs/types/user-info-request-dto.type.ts (hunk 6 above)
Comment (liza-veis) [inline packages/shared/src/modules/users/libs/types/user-info-request-dto.type.ts:3]: Rename to UserPatchRequestDto
```suggestion
type UserPatchRequestDto = {
  name: string;
};
```

File: packages/shared/src/modules/users/libs/types/user-info-response-dto.type.ts [hunk 8]
```diff
@@ -0,0 +1,6 @@
+type UserInfoResponseDto = {
+	email: string;
```
Comment (liza-veis) [inline packages/shared/src/modules/users/libs/types/user-info-response-dto.type.ts:2]: we need to return id for entities
```suggestion
type UserPatchResponseDto = {
    id: number;
	email: string;
```

File: packages/shared/src/modules/users/libs/types/user-auth-response-dto.type.ts [hunk 9]
```diff
@@ -1,8 +1,8 @@
+import { type UserInfoResponseDto } from "./user-info-response-dto.type.js";
+
 type UserAuthResponseDto = {
 	createdAt: string;
-	email: string;
 	id: number;
-	name: string;
-};
+} & UserInfoResponseDto;
```
Comment (liza-veis) [inline packages/shared/src/modules/users/libs/types/user-auth-response-dto.type.ts:6]: We need to either pick all these properties from UserGetAllItemResponseDto, including createdAt and id or leave it as it was, without combining types

File: apps/frontend/src/pages/profile/components/edit-user-form/edit-user-form.tsx [hunk 10]
```diff
@@ -0,0 +1,87 @@
...
+				void dispatch(authActions.getAuthenticatedUser());
+			})(event_);
+		},
+		[dispatch, handleSubmit, userId],
+	);
+
+	useEffect(() => {
+		const isEqual = areObjectsEqual(watchedValues, defaultValues);
+		setIsChanged(!isEqual);
+	}, [watchedValues, defaultValues]);
```
Comment (liza-veis) [inline apps/frontend/src/pages/profile/components/edit-user-form/edit-user-form.tsx:54]: Aren't there properties like isTouched or isDirty or isFormChanged in react-hook-form?

File: apps/frontend/src/modules/users/users-api.ts [hunk 11]
```diff
@@ -29,6 +33,23 @@ class UserApi extends BaseHTTPApi {
 
 		return await response.json<UserGetAllResponseDto>();
 	}
+
+	public async update(
```
Comment (liza-veis) [inline apps/frontend/src/modules/users/users-api.ts:37]: we will use update for put
```suggestion
	public async patch(
```

**Review by liza-veis — CHANGES_REQUESTED** [review]

File: apps/frontend/src/pages/profile/components/edit-user-form/edit-user-form.tsx (hunk 4 above)
Comment (liza-veis) [reply apps/frontend/src/pages/profile/components/edit-user-form/edit-user-form.tsx:36]: yep, we can't use event because we have an `event` global variable

File: apps/frontend/src/pages/profile/components/edit-user-form/edit-user-form.tsx (hunk 10 above)
Comment (perekotypole) [reply apps/frontend/src/pages/profile/components/edit-user-form/edit-user-form.tsx:54]: Oh, really! I completely missed it

File: packages/shared/src/modules/users/libs/types/user-auth-response-dto.type.ts (hunk 9 above)
Comment (perekotypole) [reply packages/shared/src/modules/users/libs/types/user-auth-response-dto.type.ts:6]: Okay, I get it. I'd rather leave it the way it was so as not to make a mess

File: apps/frontend/src/libs/components/input/input.tsx [hunk 12]
```diff
@@ -13,6 +13,7 @@ import styles from "./styles.module.css";
 type Properties<T extends FieldValues> = {
 	autoComplete?: string;
 	control: Control<T, null>;
+	disabled?: boolean;
```
Comment (what1s1ove) [inline apps/frontend/src/libs/components/input/input.tsx:16]: b19

File: apps/backend/src/modules/users/user.controller.ts [hunk 13]
```diff
@@ -41,6 +46,18 @@ class UserController extends BaseController {
 			method: "GET",
 			path: UsersApiPath.ROOT,
 		});
+
+		this.addRoute({
+			handler: (options) =>
+				this.patch(
+					options as APIHandlerOptions<{
+						body: UserPatchResponseDto;
```
Comment (GvoFor) [inline apps/backend/src/modules/users/user.controller.ts:54]:
```suggestion
						body: UserPatchRequestDto;
```

File: apps/frontend/src/modules/users/slices/actions.ts [hunk 14]
```diff
@@ -15,4 +19,17 @@ const loadAll = createAsyncThunk<
 	return userApi.getAll();
 });
 
-export { loadAll };
+const updateProfile = createAsyncThunk<
+	UserPatchResponseDto,
+	{ id: number; userPayload: UserPatchResponseDto },
```
Comment (GvoFor) [inline apps/frontend/src/modules/users/slices/actions.ts:24]:
```suggestion
	{ id: number; payload: UserPatchRequestDto },
```

File: apps/frontend/src/pages/profile/profile.tsx [hunk 15]
```diff
@@ -0,0 +1,35 @@
...
+		return <></>;
+	}
+
+	const isLoading =
+		dataStatus === DataStatus.PENDING || dataStatus === DataStatus.IDLE;
+
+	return (
+		<PageLayout>
+			<div className={styles["profile-content"]}>
+				<h1 className={styles["label"]}>Profile</h1>
```
Comment (liza-veis) [inline apps/frontend/src/pages/profile/profile.tsx:21]: should be title, it will be renamed in other places in another PR
```suggestion
				<h1 className={styles["title"]}>Profile</h1>
```

**Review by liza-veis — CHANGES_REQUESTED** [review]

**Review by what1s1ove — APPROVED** [review]

File: apps/frontend/src/libs/components/button/styles.module.css [hunk 16]
```diff
@@ -13,6 +13,11 @@
 	transition: background-color 0.3s ease;
 }
 
-.button:hover {
+.button:disabled {
+	cursor: unset;
```
Comment (liza-veis) [inline apps/frontend/src/libs/components/button/styles.module.css:17]:
```suggestion
	cursor: not-allowed;
```

**Review by liza-veis — CHANGES_REQUESTED** [review]

**Review by liza-veis — APPROVED** [review]
